snuba/query/validation/signature.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC
+from typing import Sequence, Set, Type
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.clickhouse.columns import ColumnType as ClickhouseType
+from snuba.query.expressions import Expression
+from snuba.query.matchers import Any, Column, Param
+from snuba.query.parser.exceptions import FunctionValidationException
+from snuba.query.validation import FunctionCallValidator
+
+
+class ParamType(ABC):
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        raise NotImplementedError
+
+
+class AnyType(ParamType):
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        return
+
+
+COLUMN_PATTERN = Column(
+    alias=None, table_name=None, column_name=Param("column_name", Any(str)),
+)
+
+
+class ColumnType(ParamType):
+    """
+    Validates the that the type of a Column expression is in a set
+    of allowed types.
+    """
+
+    def __init__(
+        self, types: Set[Type[ClickhouseType]], column_required: bool = False
+    ) -> None:
+        self.__valid_types = types
+        # If true, this will fail if the expression is not a simple column.
+        self.__column_required = column_required
+
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        match = COLUMN_PATTERN.match(expression)
+        if match is None:
+            if self.__column_required:
+                raise FunctionValidationException(
+                    f"Illegal expression: {expression}. Column required."
+                )
+            else:
+                return
+
+        column_name = match.string("column_name")
+        column = schema.get(column_name)
+        if column is None:
+            # TODO: We cannot raise exceptions if the column is not present
+            # on the schema just yet because the current logical schemas are
+            # sadly not complete. Fix them and then raise an exception in this
+            # case.
+            return
+
+        column_type = column.type.get_raw()
```

Comment:
Do we need to consider any column type modifiers like nullable?

Reply:
There are indeed some functions that may require it. I don't know whether we are going to write a validator for them at this point. Will try to see how it would look like that way.

Reply:
done